# Detailed table-of-contents exploration

**Open Library is our main source of table-of-contents (TOC) data.** A TOC lists a book's chapters and other sections. Our saved dataset combines Amazon book records with information from matched Open Library editions. An edition is a particular published version of a book.

We explore the 9,034 rows already downloaded from BigQuery into `data/raw/books.parquet`. Each row has an Amazon book identifier, `parent_asin`, and its matched Open Library information. More than one Amazon record can refer to the same Open Library edition, so 9,034 rows does not necessarily mean 9,034 different editions.

The aim is to understand which TOC text and details are available before deciding how to prepare them for recommendations. Every number below refers to this saved file.

Run the cells in order using the project's `.venv` kernel. This notebook reads the local file; it does not query BigQuery or contact Open Library.

## 1. Load the local snapshot

Objective: Load the saved books and record the file checksum before exploration. From `.parquet` file.

In [1]:
import hashlib
import json
import re
from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from IPython.display import display

from bookpath.local_data import read_books_from_parquet

project_root = Path.cwd()
while not (project_root / "src/bookpath").is_dir():
    if project_root == project_root.parent:
        raise FileNotFoundError("Open this notebook inside the Bookpath repository.")
    project_root = project_root.parent

raw_path = project_root / "data/raw/books.parquet"


def file_checksum(path):
    with path.open("rb") as raw_file:
        return hashlib.file_digest(raw_file, "sha256").hexdigest()


checksum_before = file_checksum(raw_path)
books = read_books_from_parquet(raw_path)
print(f"Books: {len(books):,}; checksum: {checksum_before}")

Books: 9,034; checksum: 4acf6979b15215c28f2a45a1488ace4d749e2573f0fa4a0110e2933dd0194dd8


## 2. Check the two saved copies of each book's TOC

Objective: Check whether the separate `raw_toc_json` column kept the **same TOC contents and order** as the saved `openlibrary_raw_edition_json`. Equal quantities alone would not be enough: two lists can have the same length but different chapter titles.

Both columns below belong to the `books` DataFrame loaded from `data/raw/books.parquet`.

| Column in `books` | What one cell stores | What we compare after `json.loads` |
| --- | --- | --- |
| `raw_toc_json` | A JSON string containing just that book's TOC | The entire decoded list |
| `openlibrary_raw_edition_json` | A JSON string containing the saved Open Library edition record, including its TOC | The list under the dictionary key `table_of_contents` |

`json.loads` converts JSON text into Python objects so we can access their contents. The first line below creates `raw_tocs` from `books["raw_toc_json"]`: it holds one decoded TOC list per book row. **Section 3 reuses this variable.**

**Why do this?** If the two lists agree for a book, we can inspect its separate `raw_toc_json` column without losing information from the saved edition record's `openlibrary_records = { "table_of_contents": "..." }` list. This checks the saved copies, not whether the source TOC is complete or correctly matched to the book. It does not compare the `toc_entries` column yet.

The final table answers a separate question: **what kind of items does each book's TOC list contain?** It summarizes the saved `raw_toc_shape` column. `list[dict]` means a list of dictionaries; `list[string]` means a list of plain strings. Each book record contributes once to this table.

In [2]:
# Each cell starts as a JSON string; decode it into a Python list of TOC items.
raw_tocs = books["raw_toc_json"].map(json.loads)

# Each full Open Library JSON string becomes a Python dictionary.
openlibrary_records = books["openlibrary_raw_edition_json"].map(json.loads)
# Select the original TOC list from each dictionary.
openlibrary_tocs = openlibrary_records.map(lambda record: record.get("table_of_contents"))

# Pair the two lists for the same book. Equality checks their contents and order.
# strict=True also checks that we compare the same number of books on both sides.
source_matches = pd.Series(
    [raw == original for raw, original in zip(raw_tocs, openlibrary_tocs, strict=True)],
    index=books.index,
)
print(f"Book rows with identical TOC contents and order in the two saved columns: {source_matches.sum():,} / {len(books):,}")
print("TOC-list format per book record (not the number of TOC items):")
display(books["raw_toc_shape"].value_counts().rename("book_records").to_frame())

Book rows with identical TOC contents and order in the two saved columns: 9,034 / 9,034
TOC-list format per book record (not the number of TOC items):


,book_records
raw_toc_shape,
list[dict],9030
list[string],4


**Result for this file:** All 9,034 book rows have identical TOC lists in the two saved locations, including item contents and order. We can therefore use `raw_toc_json` for the item-by-item inspection that follows.

The format table counts **book records**: 9,030 have lists of dictionaries, and 4 have lists of strings. It does not count the individual items inside those lists; section 3 does that.

Objective: Display one book's original TOC items and its parsed entries as two tables. Change `book_asin` to inspect another book by its Amazon identifier. The default is Never Split the Difference, as in docs/data.md. Original item positions and parsed sequence numbers can differ when an empty item was omitted.

`books.loc[books["parent_asin"] == book_asin]` keeps only matching rows. `.iloc[0]` then selects the first matching row as a Pandas `Series`. In this snapshot each `parent_asin` identifies one row. `selected_book["title"]` gets its book-title string; `selected_book["toc_entries"][0]["text"]` gets the first parsed TOC entry's text. `where()` would keep nonmatching rows and replace their values with missing values, so it is not the row filter we need here.

In this Parquet snapshot, `toc_entries` is a NumPy array of dictionaries. `list()` converts that array to a Python list while keeping the same dictionaries in order. Passing this list to `DataFrame` gives one column per dictionary key (`text`, `page`, etc.). Passing the array directly gives a single column containing whole dictionaries.

In [3]:
book_asin = "0062407805"
# Filter by the book ID, then select its row as a Pandas Series.
selected_book = books.loc[books["parent_asin"] == book_asin].iloc[0]
# Decode the raw TOC from that same row into a Python list.
selected_raw_toc = json.loads(selected_book["raw_toc_json"])  # COLUMN lookup: read this cell from the selected book row.
raw_view = pd.DataFrame({
    "raw_position": range(1, len(selected_raw_toc) + 1),
    "raw_item": selected_raw_toc,
})

# NumPy array of dictionaries -> Python list of dictionaries -> table of fields.
# list() changes the container for display; it does not clean any entry text.
parsed_entries = list(selected_book["toc_entries"])  # COLUMN lookup: read this cell from the selected book row.
parsed_view = pd.DataFrame(parsed_entries)

print(f"Selected parent_asin: {selected_book['parent_asin']}")
print(f"Amazon title: {selected_book['title']}")
print(f"Open Library title: {selected_book['ol_title']}")
with pd.option_context("display.max_colwidth", None):
    print("Original TOC: each raw_item cell contains an entire original item.")
    display(raw_view)
    print("Parsed TOC: each dictionary field has its own column.")
    display(parsed_view)

Selected parent_asin: 0062407805
Amazon title: Never Split the Difference: Negotiating As If Your Life Depended On It
Open Library title: Never Split the Difference
Original TOC: each raw_item cell contains an entire original item.


,raw_position,raw_item
0,1,"{'level': 0, 'label': 'Chapter 1', 'title': 'The New Rules', 'pagenum': '1', 'subtitle': 'How to Become the Smartest Person in Any Room', 'type': {'key': '/type/toc_item'}}"
1,2,"{'level': 0, 'label': 'Chapter 2', 'title': 'Be a Mirror', 'pagenum': '23', 'subtitle': 'How to Quickly Establish Rapport', 'type': {'key': '/type/toc_item'}}"
2,3,"{'level': 0, 'label': 'Chapter 3', 'title': 'Don't Feel Their Pain, Label It', 'pagenum': '49', 'subtitle': 'How to Create Trust with Tactical Empathy', 'type': {'key': '/type/toc_item'}}"
3,4,"{'level': 0, 'label': 'Chapter 4', 'title': 'Beware ""Yes""-Master ""No""', 'pagenum': '74', 'subtitle': 'How to Generate Momentum and Make It Safe to Reveal the Real Stakes', 'type': {'key': '/type/toc_item'}}"
4,5,"{'level': 0, 'label': 'Chapter 5', 'title': 'Trigger the Two Words That Immediately Transform Any Negotiation', 'pagenum': '96', 'subtitle': 'How to Gain the Permission to Persuade', 'type': {'key': '/type/toc_item'}}"
5,6,"{'level': 0, 'label': 'Chapter 6', 'title': 'Bend Their Reality', 'pagenum': '113', 'subtitle': 'How to Shape What Is Fair', 'type': {'key': '/type/toc_item'}}"
6,7,"{'level': 0, 'label': 'Chapter 7', 'title': 'Create the Illusion of Control', 'pagenum': '140', 'subtitle': 'How to Calibrate Questions to Transform Conflict into Collaboration', 'type': {'key': '/type/toc_item'}}"
7,8,"{'level': 0, 'label': 'Chapter 8', 'title': 'Guarantee Execution', 'pagenum': '162', 'subtitle': 'How to Spot the Liars and Ensure Follow-Through from Everyone Else', 'type': {'key': '/type/toc_item'}}"
8,9,"{'level': 0, 'label': 'Chapter 9', 'title': 'Bargain Hard', 'pagenum': '188', 'subtitle': 'How to Get Your Price', 'type': {'key': '/type/toc_item'}}"
9,10,"{'level': 0, 'label': 'Chapter 10', 'title': 'Find the Black Swan', 'pagenum': '213', 'subtitle': 'How to Create Breakthroughs by Revealing the Unknown Unknowns', 'type': {'key': '/type/toc_item'}}"


Parsed TOC: each dictionary field has its own column.


,level,page,sequence,source_key,text
0,0,1,1,title,The New Rules
1,0,23,2,title,Be a Mirror
2,0,49,3,title,"Don't Feel Their Pain, Label It"
3,0,74,4,title,"Beware ""Yes""-Master ""No"""
4,0,96,5,title,Trigger the Two Words That Immediately Transform Any Negotiation
5,0,113,6,title,Bend Their Reality
6,0,140,7,title,Create the Illusion of Control
7,0,162,8,title,Guarantee Execution
8,0,188,9,title,Bargain Hard
9,0,213,10,title,Find the Black Swan


## 3. What information is inside a TOC?

So far, each row has represented one book. Now we will look at individual TOC items, such as chapters, introductions, and indexes.

### 3.1 Look at one item

Start with the first item from the book selected in section 2. A dictionary stores named details; a plain string stores only text.

In [4]:
first_item = selected_raw_toc[0]
if isinstance(first_item, dict):
    display(pd.Series(first_item, name="Value").reindex([
        "label", "title", "subtitle", "pagenum", "level",
    ]).rename_axis("Field").to_frame())
else:
    print(first_item)

,Value
Field,
label,Chapter 1
title,The New Rules
subtitle,How to Become the Smartest Person in Any Room
pagenum,1
level,0


For *Never Split the Difference* (`parent_asin = "0062407805"`), this item includes a chapter label, title, subtitle, page, and level. `pagenum` is the page label; `level` records depth within the TOC.

### 3.2 Give each item its own row

Keep following this same book. **Before `explode()`, one row holds its list of 14 items. After `explode()`, 14 rows each hold one item**, with the book ID repeated alongside it. The items themselves stay the same.

The code transforms the whole catalog; the before/after tables show only our example book. Long cells are shortened for display, and the after table shows the first three of its 14 rows.

The `.loc[...]` line keeps books whose TOC list has at least one item. It does not remove individual items with empty text. `ignore_index=True` gives the resulting rows a fresh row index.

In [5]:
toc_items = books[["parent_asin"]].copy()
toc_items["raw_item"] = raw_tocs

print(f"Before: one row for {book_asin}; raw_item holds a list of {len(selected_raw_toc)} items.")
with pd.option_context("display.max_colwidth", 160):
    display(toc_items.loc[toc_items["parent_asin"] == book_asin])

toc_items = toc_items.loc[raw_tocs.map(len).gt(0)]
print("Book records excluded because their TOC list is empty:", int(raw_tocs.map(len).eq(0).sum()))
toc_items = toc_items.explode("raw_item", ignore_index=True)

book_items = toc_items.loc[toc_items["parent_asin"] == book_asin]
print(f"After: {len(book_items)} rows for {book_asin}; each raw_item holds one item. First three rows:")
with pd.option_context("display.max_colwidth", 160):
    display(book_items.head(3))
print(f"Whole catalog: {len(books):,} book records → {len(toc_items):,} TOC items.")

Before: one row for 0062407805; raw_item holds a list of 14 items.


,parent_asin,raw_item
131,0062407805,"[{'level': 0, 'label': 'Chapter 1', 'title': 'The New Rules', 'pagenum': '1', 'subtitle': 'How to Become the Smartest Person in Any Room', 'type': {'key': '..."


Book records excluded because their TOC list is empty: 0
After: 14 rows for 0062407805; each raw_item holds one item. First three rows:


,parent_asin,raw_item
1787,0062407805,"{'level': 0, 'label': 'Chapter 1', 'title': 'The New Rules', 'pagenum': '1', 'subtitle': 'How to Become the Smartest Person in Any Room', 'type': {'key': '/..."
1788,0062407805,"{'level': 0, 'label': 'Chapter 2', 'title': 'Be a Mirror', 'pagenum': '23', 'subtitle': 'How to Quickly Establish Rapport', 'type': {'key': '/type/toc_item'}}"
1789,0062407805,"{'level': 0, 'label': 'Chapter 3', 'title': 'Don't Feel Their Pain, Label It', 'pagenum': '49', 'subtitle': 'How to Create Trust with Tactical Empathy', 'ty..."


Whole catalog: 9,034 book records → 141,513 TOC items.


In [6]:
item_formats = toc_items["raw_item"].map(type).value_counts().rename(index={
    dict: "Named details (dictionary)", str: "Plain text (string)",
})
item_formats.loc["Total TOC items"] = item_formats.sum()
display(item_formats.rename_axis("Item format").rename("TOC items").to_frame())

,TOC items
Item format,
Named details (dictionary),141384
Plain text (string),129
Total TOC items,141513


**Takeaway:** Most items contain named details; only 129 are plain strings.

### 3.3 Which details are available across the books?

The **rows** list every field found inside the original TOC dictionaries, such as `title`, `authors`, and `subtitle`. The **columns** are measurements we calculate about those fields:

- **`items_with_key`**: how many TOC items contain the field.
- **`books_with_key`**: how many distinct book records contain at least one such item. This prevents a long TOC from looking like many different books.
- **`books_percent`**: that book count as a percentage of all 9,034 book records, making common and rare fields easier to compare.

The table is sorted from most to least frequent by item count. We do not add the rows together: one item can contain several fields. Presence alone does not mean the value contains text; section 4 checks that, and section 5 checks what the parsed version kept.

In [7]:
property_rows = []
for row in toc_items.itertuples():
    if isinstance(row.raw_item, dict):
        for key, value in row.raw_item.items():
            property_rows.append({
                "parent_asin": row.parent_asin, "property": key, "value": value,
            })

raw_properties = pd.DataFrame(property_rows)
property_summary = raw_properties.groupby("property").agg(
    items_with_key=("parent_asin", "size"),
    books_with_key=("parent_asin", "nunique"),
)
display(property_summary.assign(
    books_percent=(100 * property_summary["books_with_key"] / len(books)).round(2),
).sort_values("items_with_key", ascending=False))

,items_with_key,books_with_key,books_percent
property,,,
type,141384,9030,99.96
title,140446,8982,99.42
level,131126,8354,92.47
pagenum,6374,170,1.88
label,3801,204,2.26
value,773,48,0.53
class,69,5,0.06
subtitle,16,2,0.02
authors,6,5,0.06


**Reading one row:** `authors: 6, 5` means six TOC items have an `authors` field, spread across five book records. It counts items with that field, not individual authors.

**Why this matters:** `title` is available across most book records; chapter subtitles and author details are rare. `type` is most frequent but describes the stored item's format, not the chapter's subject. Frequency helps us understand availability; it does not rank fields by usefulness for recommendations. Next, we check which fields supply actual text.

## 4. Why are there fewer parsed entries?

`raw_toc_json` is a column in `books`. Each cell is JSON text; after `json.loads()`, it becomes a **list of TOC items**. We call each element in that list an **original item**. There is no nested key called “original items”. The `raw_item` column created in section 3 holds one of these elements per row.

`toc_entries` is a separate column in `books`. Its **parsed entries** store the selected chapter text under `text`. For our book's first chapter, the original item has `title = "The New Rules"`; the parsed entry has `text = "The New Rules"`.

### 4.1 Which items supply main text?

For each original item, `select_main_text()` looks for text in **`title`, then `value`, then `label`**. It uses the first one that is not empty or just spaces. The chosen words go into `main_text`; the field used (for example, `"title"`) goes into `text_field`.

Some items are just text, such as `"Introduction"`, instead of a dictionary like `{"title": "Introduction"}`. We use that text directly and call it **plain string** in the table. No field was needed, so `text_field` is `None` for those items. An item such as `{"value": ""}` supplies no text.

The examples above illustrate the possible formats. The counts below cover the whole catalog and include a subtotal for items with text and a total for all original items.

In [ ]:
def select_main_text(item):
    if isinstance(item, str):
        return item, None
    for field in ["title", "value", "label"]:
        value = item.get(field)
        if isinstance(value, str) and value.strip():
            return value, field
    return None, None


toc_items[["main_text", "text_field"]] = pd.DataFrame(
    toc_items["raw_item"].map(select_main_text).tolist(), index=toc_items.index,
)
toc_items["has_text"] = toc_items["main_text"].fillna("").str.strip().ne("")
items_without_main_text = toc_items.loc[~toc_items["has_text"]]

text_source_counts = (
    toc_items.loc[toc_items["has_text"], "text_field"].fillna("plain string").value_counts()
)
text_source_counts.loc["Subtotal: items with main text"] = text_source_counts.sum()
text_source_counts.loc["no main text"] = len(items_without_main_text)
text_source_counts.loc["Total original items"] = len(toc_items)
display(text_source_counts.rename_axis("Text source").rename("TOC items").to_frame())

**Answer:** The four text sources add up to **141,495 items with main text**. Adding the **18 without text** gives **141,513 original items**, as shown in the table. The next check compares this with the number of entries already saved in `toc_entries`.

### 4.2 Do the items without text explain the difference?

Compare the two totals, then check each book separately. If the explanation holds, every book's gap should equal its number of items without main text.

In [ ]:
toc_counts = books[["parent_asin", "title"]].copy()
toc_counts["actual_raw_count"] = raw_tocs.map(len)
toc_counts["actual_parsed_count"] = books["toc_entries"].map(len)
toc_counts["items_without_text"] = (
    toc_counts["parent_asin"].map(
        items_without_main_text.groupby("parent_asin").size()
    ).fillna(0).astype(int)
)
toc_counts["raw_minus_parsed"] = (
    toc_counts["actual_raw_count"] - toc_counts["actual_parsed_count"]
)
display(pd.Series({
    "Original items": len(toc_items),
    "Parsed entries": toc_counts["actual_parsed_count"].sum(),
    "Difference (original minus parsed)": toc_counts["raw_minus_parsed"].sum(),
    "Items without main text": len(items_without_main_text),
    "Book records containing these items": items_without_main_text["parent_asin"].nunique(),
    "Books with an unexplained count difference": toc_counts["raw_minus_parsed"].ne(
        toc_counts["items_without_text"]
    ).sum(),
}, name="Count").to_frame())

**Answer: yes.** **141,513 original items − 141,495 parsed entries = 18 items.** Those 18 have no usable main text and are spread across 17 book records. The explanation also holds for each book: there are zero unexplained count differences.

**There are fewer parsed entries because the earlier extraction omitted items without main text.** Section 5 checks that the remaining text and its order agree; matching counts alone cannot prove that.

### 4.3 Does our example book have fewer parsed entries?

Keep using *Never Split the Difference* (`0062407805`). Compare its own counts with the whole-catalog result above:

In [ ]:
print(f"Book: {selected_book['title']} ({book_asin})")
display(toc_counts.loc[toc_counts["parent_asin"] == book_asin, [
    "parent_asin", "actual_raw_count", "actual_parsed_count",
    "items_without_text", "raw_minus_parsed",
]])

**Answer: no.** This book has **14 original items and 14 parsed entries**, with no items lacking main text. The 18-item difference belongs to 17 other book records. Keeping this example separate from the catalog totals shows that the difference does not affect every book.

### 4.4 Are the recorded counts correct?

`raw_toc_item_count` and `toc_entry_count` store each book's counts. For each column, compare the **sum of the saved counts** with the **total we measured from the lists**, and check every book individually. Totals alone could hide errors that cancel each other out.

In [ ]:
count_checks = []
for saved_column, measured_column in [
    ("raw_toc_item_count", "actual_raw_count"),
    ("toc_entry_count", "actual_parsed_count"),
]:
    saved_counts = pd.to_numeric(books[saved_column])
    measured_counts = toc_counts[measured_column]
    count_checks.append({
        "Saved column": saved_column,
        "Saved total": int(saved_counts.sum()),
        "Measured total": int(measured_counts.sum()),
        "Book records checked": len(books),
        "Book records with an incorrect count": int(saved_counts.ne(measured_counts).sum()),
    })
display(pd.DataFrame(count_checks).set_index("Saved column"))

**Answer: yes.** The saved and measured totals agree: **141,513 original items** and **141,495 parsed entries**. All **9,034 book records** pass both checks, with zero incorrect counts. The difference between versions comes from which items supply text, not a counting error.

## 5. What did the parsed version keep?

### 5.1 Was the selected text preserved?

First, give each saved parsed entry its own row, as we did for original items. `json_normalize()` turns each dictionary's fields into columns; `join()` attaches the book ID beside them.

In [ ]:
parsed_items = books[["parent_asin", "toc_entries"]].copy()
parsed_items = parsed_items.loc[parsed_items["toc_entries"].map(len).gt(0)]
parsed_items = parsed_items.explode("toc_entries", ignore_index=True)
parsed_items = parsed_items[["parent_asin"]].join(
    pd.json_normalize(parsed_items["toc_entries"])
)
# cumcount() numbers entries within each book, starting at zero; add one.
parsed_items["parsed_position"] = parsed_items.groupby("parent_asin").cumcount() + 1
parsed_source_counts = parsed_items["source_key"].fillna("Plain string").value_counts()
parsed_source_counts.loc["Total parsed entries"] = parsed_source_counts.sum()
display(parsed_source_counts.rename_axis("Selected text field").rename("Entries").to_frame())

Pair each original item that has main text with the next parsed entry **in the same book**. Compare text, selected field, level, and page. Original text is trimmed at its ends; numeric metadata is compared as text, so `0` and `"0"` agree.

`merge()` pairs rows by book ID and position. Keeping unmatched rows makes missing entries visible.

In [ ]:
toc_items["raw_position"] = toc_items.groupby("parent_asin").cumcount() + 1
usable_items = toc_items.loc[toc_items["has_text"]].copy()
usable_items["parsed_position"] = usable_items.groupby("parent_asin").cumcount() + 1
preservation_review = usable_items.merge(
    parsed_items, on=["parent_asin", "parsed_position"], how="outer",
    indicator=True, validate="one_to_one",
)
preservation_review["text_matches"] = (
    preservation_review["main_text"].astype("string").str.strip()
    .eq(preservation_review["text"].astype("string")).fillna(False)
)
preservation_review["source_key_matches"] = (
    preservation_review["text_field"].fillna("<missing>")
    .eq(preservation_review["source_key"].fillna("<missing>"))
)

In [ ]:
def raw_field_as_text(item, field):
    value = item.get(field) if isinstance(item, dict) else None
    return None if value is None else str(value)


for raw_field, parsed_field in [("level", "level"), ("pagenum", "page")]:
    raw_values = preservation_review["raw_item"].map(
        lambda item: raw_field_as_text(item, raw_field)
    ).astype("string")
    parsed_values = preservation_review[parsed_field].astype("string")
    preservation_review[f"{parsed_field}_matches"] = (
        raw_values.fillna("<missing>").eq(parsed_values.fillna("<missing>"))
    )

comparison_fields = ["text_matches", "source_key_matches", "level_matches", "page_matches"]
comparison_summary = (~preservation_review[comparison_fields]).sum().astype(int)
comparison_summary.index = ["Main text", "Selected field", "Level", "Page"]
comparison_summary.loc["Unpaired items or entries"] = preservation_review["_merge"].ne("both").sum()
display(comparison_summary.rename("Disagreements").to_frame())

**Takeaway:** All 141,495 items with main text match the saved parsed entries in order for these four fields. This checks the extraction, not the completeness or edition match of the source TOC.

### 5.2 Does matching the main text mean every detail was kept?

**What we are checking:** Whether reading only the parsed entry's `text` gives us all the chapter information in the original item. Section 5.1 checked the selected text, level, and page; it did not check extra details such as chapter labels and subtitles.

Compare the same first chapter of *Never Split the Difference* (`0062407805`) below. The left column reads its original item; the right reads its saved parsed entry. **—** means that field is absent or has no value.

In [ ]:
example_fields = first_item if isinstance(first_item, dict) else {}
example_parsed = selected_book["toc_entries"][0]
display(pd.DataFrame({
    "Detail": ["Main text", "Chapter label", "Chapter subtitle"],
    "Original item": [select_main_text(first_item)[0], example_fields.get("label"),
                      example_fields.get("subtitle")],
    "Parsed entry": [example_parsed.get("text"), example_parsed.get("label"),
                     example_parsed.get("subtitle")],
}).set_index("Detail").fillna("—"))

**Answer: no.** For this chapter, both versions contain “The New Rules”, but the parsed `text` leaves out **“Chapter 1”** and **“How to Become the Smartest Person in Any Room”**. There are no separate label or subtitle fields in this parsed entry either. Those details remain available in `raw_toc_json`.

**What this establishes:** Matching the main text does not mean the parsed version kept every detail. If we want this chapter's label or subtitle, we must read the original item. This example does not establish that adding those details improves recommendations; that requires evaluation. Other items may already use a label as their main text.

### 5.3 Which parsed fields are missing?

Count absent and blank values separately. Percentages use all parsed entries. Entries that came from plain text have no selected field name, so a missing `source_key` is expected for them.

In [ ]:
coverage_rows = []
for field in ["text", "sequence", "level", "page", "source_key"]:
    values = parsed_items[field].astype("string")
    coverage_rows.append({
        "Field": field,
        "Missing": int(values.isna().sum()),
        "Blank": int(values.str.strip().eq("").sum()),
        "Missing (%)": round(100 * values.isna().mean(), 2),
    })
display(pd.DataFrame(coverage_rows).set_index("Field"))

**Takeaway:** Parsed text is present throughout, but pages are missing for 95.50% of entries and levels for 7.33%.

## 6. Can we follow the order and structure?

### 6.1 Are entries numbered in order?

Compare saved sequence numbers with the actual position of each parsed entry within its book.

In [ ]:
sequence_values = pd.to_numeric(parsed_items["sequence"], errors="coerce")
sequence_disagreements = sequence_values.ne(parsed_items["parsed_position"])
print("Entries whose sequence differs from their position:", int(sequence_disagreements.sum()))

### 6.2 Do levels describe chapters and subsections?

A level records depth, such as a section within a chapter. Count the saved levels and flag jumps larger than one level for inspection.

In [ ]:
display(parsed_items["level"].fillna("Missing").value_counts()
        .rename_axis("Level").rename("Entries").to_frame())
level_values = pd.to_numeric(parsed_items["level"], errors="coerce")
level_steps = level_values.groupby(parsed_items["parent_asin"]).diff()
large_level_jumps = level_steps.gt(1)
levels_per_book = parsed_items.groupby("parent_asin")["level"].nunique()

print("Book records with multiple levels:", int(levels_per_book.gt(1).sum()))
print("Book records without levels:", int(levels_per_book.eq(0).sum()))
print("Entries with a level jump larger than one:", int(large_level_jumps.sum()))
display(parsed_items.loc[large_level_jumps, [
    "parent_asin", "parsed_position", "text", "level",
]].head(3))

**Takeaway:** Only 97 book records have multiple recorded levels. A flat or missing level does not prove the book has no subsections.

### 6.3 Are page labels always numbers?

Look at values beyond plain digits. Roman numerals and ranges can be valid page labels.

In [ ]:
page_values = parsed_items["page"].astype("string")
present_pages = page_values.notna() & page_values.str.strip().ne("")
integer_pages = page_values.str.fullmatch(r"[0-9]+").fillna(False)
other_pages = present_pages & ~integer_pages
print("Entries with a page:", int(present_pages.sum()))
print("Entries with a page label beyond plain digits:", int(other_pages.sum()))
display(page_values.loc[other_pages].value_counts().head(5)
        .rename_axis("Page label").rename("Entries").to_frame())

## 7. Does the text need closer inspection?

### 7.1 Which entries look unusual?

Flag missing text, markup, unusual characters, and numeric-only text. Checks can overlap; a flag asks for review and does not remove anything.

In [ ]:
entry_text = parsed_items["text"].astype("string")
text_flags = pd.DataFrame({
    "missing_text": entry_text.isna(),
    "blank_text": entry_text.str.strip().eq(""),
    "surrounding_whitespace": entry_text.ne(entry_text.str.strip()),
    "html_like_markup": entry_text.str.contains(r"<[^>]+>", regex=True),
    "replacement_character": entry_text.str.contains("\ufffd", regex=False),
    "control_characters": entry_text.str.contains(r"[\x00-\x08\x0b\x0c\x0e-\x1f]", regex=True),
    "numeric_only": entry_text.str.fullmatch(r"\s*[0-9]+\s*"),
    "line_breaks": entry_text.str.contains(r"[\r\n]", regex=True),
}).fillna(False)
flag_rows = []
for flag in text_flags.columns:
    flagged = text_flags[flag]
    flag_rows.append({
        "check": flag,
        "entries": int(flagged.sum()),
        "books": parsed_items.loc[flagged, "parent_asin"].nunique(),
    })
display(pd.DataFrame(flag_rows).set_index("check").rename(
    index=lambda label: label.replace("_", " "),
    columns={"entries": "Entries", "books": "Book records"},
))
display(parsed_items.loc[text_flags.any(axis=1), ["parent_asin", "parsed_position", "text"]].head(3))

**Takeaway:** Numeric-only text and markup occur in a small number of records. Inspect their context before deciding whether to clean them.

### 7.2 Are headings repeated within a book?

Count exact repeats, then repeat the check ignoring case and extra spaces. Counts include every occurrence, including the first. Repeated headings can be legitimate.

In [ ]:
duplicate_review = parsed_items[["parent_asin", "parsed_position", "text"]].copy()
duplicate_review["comparison_text"] = entry_text.str.casefold().str.replace(r"\s+", " ", regex=True).str.strip()
exact_repeats = duplicate_review.duplicated(["parent_asin", "text"], keep=False)
normalized_repeats = duplicate_review.duplicated(["parent_asin", "comparison_text"], keep=False)
for name, repeated in [("Exact repeated text", exact_repeats), ("Case/whitespace-normalized repeated text", normalized_repeats)]:
    print(f"{name}: {repeated.sum():,} entries in {duplicate_review.loc[repeated, 'parent_asin'].nunique():,} books")
display(duplicate_review.loc[normalized_repeats, ["parent_asin", "parsed_position", "text"]].head(4))

### 7.3 How much text does a TOC contain?

Measure characters and words per entry and per book. Words are counted by spaces; these are not model token counts. Book totals exclude separators between entries. The 90th percentile means nine out of ten values are at or below that length.

In [ ]:
text_lengths = parsed_items[["parent_asin"]].copy()
text_lengths["characters"] = entry_text.str.len()
text_lengths["whitespace_words"] = entry_text.str.split().str.len()
book_text_lengths = text_lengths.groupby("parent_asin")[["characters", "whitespace_words"]].sum()
length_summary = pd.concat({
    "Per entry": text_lengths[["characters", "whitespace_words"]].describe(percentiles=[0.5, 0.9, 0.99]).T,
    "Per book": book_text_lengths.describe(percentiles=[0.5, 0.9, 0.99]).T,
})
display(length_summary[["50%", "90%", "99%", "max"]].rename(
    index={"characters": "Characters", "whitespace_words": "Words"},
    columns={"50%": "Median", "90%": "90th percentile", "99%": "99th percentile", "max": "Longest"},
))

fig, axes = plt.subplots(1, 2, figsize=(10, 3))
toc_counts["actual_parsed_count"].hist(bins=40, ax=axes[0])
axes[0].set(xlabel="Parsed entries per book", ylabel="Books")
text_lengths["characters"].hist(bins=40, ax=axes[1])
axes[1].set(xlabel="Characters per entry", ylabel="Entries")
plt.tight_layout()
plt.show()
display(books.loc[books["parent_asin"].isin(book_text_lengths.nlargest(3, "characters").index),
                  ["parent_asin", "title", "toc_entry_count"]])

**Takeaway:** The median entry has 30 characters; the median book total is 410. The longest TOCs need separate inspection before deciding how to split text for recommendations.

## 8. What does the TOC describe?

### 8.1 Chapters, volumes, or supporting material?

Look for English labels such as “Volume”, “Part”, and “Index”. These limited patterns supply examples; they do not classify every item or decide which entries to keep.

In [ ]:
content_patterns = {
    "volume_marker": r"^\s*(?:v\.|vol\.?|volume)\s+[0-9ivxlcdm]+\b",
    "part_marker": r"^\s*part\s+[0-9ivxlcdm]+\b",
    "chapter_marker": r"^\s*(?:chapter|ch\.)\s+[0-9ivxlcdm]+\b",
    "front_or_back_matter": r"^\s*(?:introduction|preface|foreword|acknowledg\w*|bibliography|index|appendix|appendices)\b",
}
for label, pattern in content_patterns.items():
    matches = entry_text.str.contains(pattern, case=False, regex=True, na=False)
    readable_label = label.replace("_", " ")
    print(f"{readable_label}: {matches.sum():,} entries; {parsed_items.loc[matches, 'parent_asin'].nunique():,} books")
    display(parsed_items.loc[matches, ["parent_asin", "parsed_position", "text"]].head(2))

### 8.2 Is there more TOC text elsewhere?

Search saved descriptions, features, and notes for headings such as “Table of Contents”. Show one excerpt per matching field. Some matches only advertise a TOC, and differently worded chapter lists may be missed.

The optional appendix separately checks nested field names for structured TOCs.

In [ ]:
contents_heading_pattern = r"\b(?:table\s+of\s+contents|list\s+of\s+chapters|contents\s*:)"
text_candidates = []
for position, book in enumerate(books.itertuples()):
    edition = openlibrary_records.iloc[position]
    fields = {
        "amazon_description": list(book.description),
        "amazon_features": list(book.features),
        "openlibrary_description": edition.get("description"),
        "openlibrary_notes": edition.get("notes"),
    }
    for field, value in fields.items():
        text = json.dumps(value, ensure_ascii=False)
        match = re.search(contents_heading_pattern, text, re.IGNORECASE)
        if match:
            text_candidates.append({
                "row_position": position,
                "parent_asin": book.parent_asin,
                "field": field,
                "excerpt": text[max(0, match.start() - 60):match.end() + 240],
            })
hidden_text_review = pd.DataFrame(text_candidates, columns=["row_position", "parent_asin", "field", "excerpt"])
display(hidden_text_review.groupby("field").size().rename("Candidate book records").to_frame())
with pd.option_context("display.max_colwidth", None):
    display(hidden_text_review.groupby("field", sort=True).head(1)[
        ["parent_asin", "field", "excerpt"]
    ])

**Takeaway:** The search finds candidates in Amazon descriptions and feature lists, and in Open Library notes. Read the full values before treating them as chapter lists; a book can appear in more than one group.

## 9. What have we learned?

- The two saved original TOC copies agree. Most items contain named details.
- Eighteen items lack main text; the remaining 141,495 agree with the parsed version for text, selected field, level, and page, in order.
- Original items retain extra details such as chapter subtitles. Parsed text alone does not include every original detail.
- Missing metadata, repeated headings, and unusual text need context before cleaning.

**Next questions:** Does each TOC match the intended edition? Which extra details help recommendations? Do apparent repeats and text-search matches contain useful chapter information?

Define retrieval evaluation before choosing cleaning or text-splitting rules. The separate feature-building step is illustrated in [notebook 03](03_toc_features_preview.ipynb); detailed findings are in [docs/data.md](../docs/data.md).

Finally, verify that this exploration left the raw file unchanged:

In [ ]:
assert file_checksum(raw_path) == checksum_before, "The raw Parquet changed during exploration."
print("Raw file unchanged.")

## Appendix: search nested field names

This optional technical check searches the saved Amazon and Open Library records for field names containing “toc”, “contents”, “chapter”, or “outline”. The function visits dictionaries and lists inside other dictionaries and lists; this is called recursion.

The table counts book records with each matching path. A matching name is a lead for inspection, not proof that its value contains a TOC. The main analysis above does not depend on this appendix.

In [ ]:
def contents_key_paths(value, path=""):
    paths = set()
    if isinstance(value, dict):
        for key, child in value.items():
            child_path = f"{path}.{key}" if path else key
            if re.search(r"toc|contents?|chapters?|outline", key, re.IGNORECASE):
                paths.add(child_path)
            paths.update(contents_key_paths(child, child_path))
    elif isinstance(value, list):
        for child in value:
            paths.update(contents_key_paths(child, path + "[]"))
    return paths


key_path_rows = []
for column in ["amazon_raw_record_json", "openlibrary_raw_edition_json"]:
    for parent_asin, record in zip(books["parent_asin"], books[column].map(json.loads), strict=True):
        for path in contents_key_paths(record):
            key_path_rows.append({"source": column, "path": path, "parent_asin": parent_asin})
key_path_summary = pd.DataFrame(key_path_rows, columns=["source", "path", "parent_asin"])
display(key_path_summary.groupby(["source", "path"]).size().rename("books").to_frame())

**Takeaway for this snapshot:** This field-name search finds Open Library's `table_of_contents` and no matching Amazon field names. The text search in section 8 can still find chapter lists inside descriptions.

### Optional examples: extra details in other book records

Section 5.2 followed only `0062407805`. For a broader inspection, this table shows one original value for each extra field found across the catalog. The book IDs identify the records to inspect next; these examples do not show whether adding the values improves recommendations.

In [ ]:
additional_properties = raw_properties.loc[
    raw_properties["property"].isin(["label", "subtitle", "authors", "description", "class"])
]
with pd.option_context("display.max_colwidth", 100):
    display(additional_properties.groupby("property", sort=True).head(1))